src/link_ops.rs

```diff
@@ -0,0 +1,39 @@
+// Copyright 2020 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+/// Base trait for link operations.
+///
+/// `LinkPtr` is the representation of a link pointer.
+/// Typically this is `NonNull`, but compact representations such
+/// as `u8` or `u16` are possible.
+pub trait LinkOps {
```

Comment:
This needs to be an unsafe trait, since we rely on `is_linked` to be correctly implemented for soundness. Also the implementer must ensure that none of the methods in `LinkOps` (and any collection-specific subtraits) panic since this could lead to unsoundness.